### PR TITLE
Accept underlying bids from Blur

### PR DESCRIPTION
**Pull request by zhoujia6139**
## Security Checklist

- [ ] 1. Re-Entrancy
- [ ] 2. Arithmetic Over/Under Flows
- [ ] 3. Unexpected Ether
- [ ] 4. Delegatecall
- [ ] 5. Default Visibilities
- [ ] 6. Entropy Illusion
- [ ] 7. External Contract Referencing
- [ ] 8. Short Address/Parameter Attack (off chain)
- [ ] 9. Unchecked CALL Return Values
- [ ] 10. Race Conditions / Front Running
- [ ] 11. Denial Of Service (DOS)
- [ ] 12. Block Timestamp Manipulation
- [ ] 13. Constructors with Care
- [ ] 14. Uninitialized Storage Pointers
- [ ] 15. Floating Points and Precision
- [ ] 16. Tx.Origin Authentication
- [ ] 17. Address.isContract Re-Entrancy via Constructor

### ⚠️ NOTES ⚠️

Make sure to think about each of these exploits in this PR.

### CODE REVIEW

File: contracts/protocol/libraries/logic/PoolExtendedLogic.sol [hunk 1]
```diff
@@ -428,15 +732,27 @@ library PoolExtendedLogic {
...
+                tokenIds,
+                timeLockParams
+            );
+        if (oldCollateralizedBalance > 0 && collateralizedBalance == 0) {
             userConfig.setUsingAsCollateral(reserveIndex, false);
             emit ReserveUsedAsCollateralDisabled(asset, user);
         }
     }
+
+    function _ensureApeCoinPositionNotExistedOn(
```
Comment (billyz9611) [inline contracts/protocol/libraries/logic/PoolExtendedLogic.sol:750]: should we also need to check BAYC/MAYC is not Paired ?

File: contracts/protocol/libraries/logic/ValidationLogic.sol [hunk 2]
```diff
@@ -55,13 +55,55 @@ library ValidationLogic {
...
+        );
+        DataTypes.ReserveConfigurationMap memory nftReserveConfiguration = ps
+            ._reserves[request.collection]
+            .configuration;
+        (, uint256 nftLiquidationThreshold, , , ) = nftReserveConfiguration
+            .getParams();
+        DataTypes.ReserveConfigurationMap memory wethReserveConfiguration = ps
+            ._reserves[weth]
+            .configuration;
+        (, uint256 wethLiquidationThreshold, , , ) = wethReserveConfiguration
```
Comment (billyz9611) [inline contracts/protocol/libraries/logic/ValidationLogic.sol:89]: should we cache wethLiquidationThreshold to reduce read times?

File: contracts/protocol/libraries/logic/ValidationLogic.sol [hunk 3]
```diff
@@ -55,13 +55,51 @@ library ValidationLogic {
      */
     uint256 public constant HEALTH_FACTOR_LIQUIDATION_THRESHOLD = 1e18;
 
-    function validateStatusForBlurExchangeRequest(
+    function validateInitiateAcceptBlurBidsRequest(
```
Comment (billyz9611) [inline contracts/protocol/libraries/logic/ValidationLogic.sol:58]: should we also need to check the NFT is not in auction ?

File: contracts/protocol/libraries/logic/ValidationLogic.sol [hunk 4]
```diff
@@ -55,13 +55,51 @@ library ValidationLogic {
...
+        address weth,
+        address oracle,
+        uint256 wethLiquidationThreshold
+    ) internal view {
+        require(
+            ps._acceptBlurBidsRequestStatus[requestHash] ==
+                DataTypes.AcceptBlurBidsRequestStatus.Default,
+            Errors.INVALID_REQUEST_STATUS
+        );
+        require(msg.sender == request.initiator, Errors.CALLER_NOT_INITIATOR);
```
Comment (billyz9611) [inline contracts/protocol/libraries/logic/ValidationLogic.sol:72]: we also should check NFT owner is msg.sender

File: contracts/protocol/libraries/logic/PoolExtendedLogic.sol [hunk 5]
```diff
@@ -317,6 +363,272 @@ library PoolExtendedLogic {
...
+                    currentOwner == nTokenOwner,
+                    Errors.NOT_SAME_NTOKEN_OWNER
+                );
+            }
+
+            // calculate and accumulate weth
+            totalETH += (request.bidingPrice - request.marketPlaceFee);
+
+            // update request status
+            delete ps._blurExchangeRequestStatus[requestHash];
```
Comment (0xrjman) [inline contracts/protocol/libraries/logic/PoolExtendedLogic.sol:539]: shoule be `delete ps._acceptBlurBidsRequestStatus[requestHash];`

File: contracts/protocol/libraries/logic/PoolExtendedLogic.sol [hunk 6]
```diff
@@ -317,6 +363,272 @@ library PoolExtendedLogic {
...
+            // check request status
+            bytes32 requestHash = _calculateAcceptBlurBidsRequestHash(request);
+            require(
+                ps._acceptBlurBidsRequestStatus[requestHash] ==
+                    DataTypes.AcceptBlurBidsRequestStatus.Initiated,
+                Errors.INVALID_REQUEST_STATUS
+            );
+
+            // update request status
+            delete ps._blurExchangeRequestStatus[requestHash];
```
Comment (0xrjman) [inline contracts/protocol/libraries/logic/PoolExtendedLogic.sol:596]: _acceptBlurBidsRequestStatus

**Pull request merged**